mysys/mf_keycache.c

```diff
@@ -3412,25 +3412,35 @@ int simple_key_cache_write(SIMPLE_KEY_CACHE_CB *keycache,
         reading the file block. If the read completes after us, it
         overwrites our new contents with the old contents. So we have to
         wait for the other thread to complete the read of this block.
-        read_block() takes care for the wait.
+        read_block_primary|secondary() takes care for the wait.
       */
-      if (!(block->status & BLOCK_ERROR) &&
-          ((page_st == PAGE_TO_BE_READ &&
-            (offset || read_length < keycache->key_cache_block_size)) ||
-           (page_st == PAGE_WAIT_TO_BE_READ)))
+      if (!(block->status & BLOCK_ERROR))
       {
-        read_block(keycache, block,
-                   offset + read_length >= keycache->key_cache_block_size?
-                   offset : keycache->key_cache_block_size,
-                   offset, (page_st == PAGE_TO_BE_READ));
-        DBUG_ASSERT(keycache->can_be_used);
-        DBUG_ASSERT(block->status & (BLOCK_READ | BLOCK_IN_USE));
-        /*
-          Prevent block from flushing and from being selected for to be
-          freed. This must be set when we release the cache_lock.
-          Here we set it in case we could not set it above.
-        */
-        block->status|= BLOCK_FOR_UPDATE;
+        if (page_st == PAGE_TO_BE_READ &&
+            (offset || read_length < keycache->key_cache_block_size))
+        {
+          read_block_primary(keycache, block,
+                             offset + read_length >= keycache->key_cache_block_size?
+                             offset : keycache->key_cache_block_size,
+                             offset);
+          /*
+            Prevent block from flushing and from being selected for to be
+            freed. This must be set when we release the cache_lock.
+            Here we set it in case we could not set it above.
+          */
+          block->status|= BLOCK_FOR_UPDATE;
+
+          DBUG_ASSERT(keycache->can_be_used);
+          DBUG_ASSERT(block->status & (BLOCK_READ | BLOCK_IN_USE));
+        }
+        else if (page_st == PAGE_WAIT_TO_BE_READ)
+        {
+          read_block_secondary(keycache, block);
+          block->status|= BLOCK_FOR_UPDATE;
+
+          DBUG_ASSERT(keycache->can_be_used);
+          DBUG_ASSERT(block->status & (BLOCK_READ | BLOCK_IN_USE));
```

Comment:
These 2 asserts always accompany ```read_block_primary()``` and ```read_block_secondary()```. Could you move them into these functions?

Reply:
Well, probably last assert is only relevant for secondary request. Because primary may return BLOCK_ERROR. Looks like we had a bug here.